kubernetes/scheduler/controller.go

```diff
@@ -0,0 +1,668 @@
+	err := func(obj interface{}) error {
+		// Call done so workqueue knows we have finished processing this item
+		// Must also call forget if we don't want the item re-queued.
+		defer c.workqueue.Done(obj)
+		var key string
+		var ok bool
+		// strings come off workqueue: namespace/name. delayed nature of wq
+		// means items in informer cache may be more up to date than the item
+		// initially put in the wq.
+		if key, ok = obj.(string); !ok {
+			// As item in workqueue is invalid, call forget else would loop
+			// attempting to process an invalid work item.
+			c.workqueue.Forget(obj)
+			utilruntime.HandleError(fmt.Errorf("expected string in workqueue but got %#v", obj))
+			return nil
+		}
+
+		// Run syncHandler, passing it the namespace/name key of the resource to
+		// be synced
+		err := c.processItem(key)
+		// handleErr will handle adding to the queue again.
+		c.handleErr(err, key)
+		// Finally, if no error occurs forget the item so it's not queued again
+		c.workqueue.Forget(obj)
+		return nil
+
+	}(obj)
+
+	if err != nil {
+		utilruntime.HandleError(err)
+		return true
+	}
+	return true
+}
+
+func (c *Controller) processItem(key string) error {
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return nil
+	}
+
+	// Check if it's something without a namespace. If it is, it must be a Node.
+	if len(namespace) == 0 {
+
+		// Currently nodes are the only thing without a namespace listened for.
+		node, err := c.nodeLister.Get(name)
+
+		if err != nil {
+			// The Node  may no longer exist, in which case we stop processing.
+
+			if errors.IsNotFound(err) {
+				utilruntime.HandleError(fmt.Errorf("node '%s' in work queue no longer exists", key))
+				return nil
+			}
+
+			return err
+		}
+
+		// Pass the node to processNode
+		err = c.processNode(node)
+		return err
+	}
+
+	// It has a namespace, it must be a pod.
+	pod, err := c.podLister.Pods(namespace).Get(name)
+
+	if err != nil {
+		// The pod  may no longer exist, in which case we stop processing.
+		if errors.IsNotFound(err) {
+			utilruntime.HandleError(fmt.Errorf("pod '%s' in work queue no longer exists", key))
+			return nil
+		}
+
+		return err
+	}
+
+	// Pass the pod to processPod
+	err = c.processPod(pod)
+
+	return err
+}
+
+// processPod implements the logic for how to react to observing a pod in a
+// given state. It currently copies the files passed in at runtime to any pod
+// with a waiting initcontainer, and returns errors status updates and logs via
+// the pod's ErrChan that is stored in c.podAliveMap.
+func (c *Controller) processPod(pod *corev1.Pod) error {
+	c.Debug("processPod called", "pod", pod.ObjectMeta.Name)
+	// Assume only 1 init container Given we control this (Deploy() / Spawn()
+	// this is ok)
+	if len(pod.Status.InitContainerStatuses) != 0 {
```

Comment:
Why not make the assumption explicit with == 1?

Reply:
This is checking the init container status, which there will be >=1 (state changes are recorded as init container statuses if I recall correctly). This was to ensure I didn't try and operate on an init container that was not initialised yet... (The API would throw an error that would not recover)

Reply:
(I could make it >=1 ? )